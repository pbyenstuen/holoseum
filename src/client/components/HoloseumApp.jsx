import React, { useEffect } from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import AOS from "aos";
import useLoader from "../hooks/useLoader";
import PublicRoute from "../hocs/PublicRoute";
import ProtectedRoute from "../hocs/ProtectedRoute";
import HomePage from "./landing-page/HomePage";
import ContactPage from "./landing-page/ContactPage";
import HologramView from "./hologram-view/HologramView";
import AdminPage from "./admin/AdminPage";
import Header from "./shared/Header";
import Footer from "./shared/Footer";
import NotFound from "./NotFound";

const App = ({ api }) => {
    const { data: user, loading, reload: updateUser } = useLoader(async () => await api.auth.getUser());

    useEffect(() => {
        AOS.init({
            once: true,
            duration: 1500
        });
    }, []);

    const LandingLayout = ({ children }) => {
        return (
            <>
                <Header
                    api={api}
                    user={user}
                    updateUser={updateUser} />
                {children}
                <Footer />
            </>
        )
    }

    return (
        <Router>
            <Switch>
                <PublicRoute
                    path={"/hologrammer/:name"}
                    component={HologramView}
                    api={api} />
                <PublicRoute
                    exact path={"/"}
                    component={HomePage}
                    layout={LandingLayout} />
                <PublicRoute
                    path={"/kontakt"}
                    component={ContactPage}
                    layout={LandingLayout} />
                <ProtectedRoute
                    path={"/admin"}
                    component={AdminPage}
                    layout={LandingLayout}
                    user={user}
                    loading={loading}
                    updateUser={updateUser}
                    api={api} />
                <Route
                    path={"*"}
                    component={NotFound} />
            </Switch >
        </Router >
    )
}

export default App;