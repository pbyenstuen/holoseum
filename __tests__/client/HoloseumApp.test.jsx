import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import HoloseumApp from "../../src/client/components/HoloseumApp";

require('react-router-dom').BrowserRouter = ({ children }) => <div>{children}</div>

let container;

beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation(query => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        }))
    });
});

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component, entries) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter initialEntries={[entries]}>{component}</MemoryRouter>, container);
    });
}

describe("HoloseumApp", () => {
    it("routes to NotFound component on invalid path", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "tidvis",
                })
            }
        }

        await mountComponent(<HoloseumApp api={api} />, "/nothing");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual("404");
    });

    it("can route to correct path", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "tidvis",
                })
            }
        }

        await mountComponent(<HoloseumApp api={api} />, "/");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual("HOLOSEUM");
    });

    it("renders login component on attempt to access protected route unauthenticated", async () => {
        const api = {
            auth: {
                getUser: () => {
                    return undefined;
                }
            }
        }

        await mountComponent(<HoloseumApp api={api} />, "/admin");

        expect(container.querySelector("h2").textContent).toEqual("Admin - Logg Inn");
    });

    it("gives access to protected route if authenticated", async () => {
        const api = {
            auth: {
                getUser: () => {
                    return "tidvis";
                }
            }
        }

        await mountComponent(<HoloseumApp api={api} />, "/admin");

        expect(container.querySelector("h2").textContent).toEqual("Last opp hologram");
    });
});