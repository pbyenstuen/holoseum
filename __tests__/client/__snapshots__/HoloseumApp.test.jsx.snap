// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoloseumApp can route to correct path 1`] = `"<div><header><nav><ul><li><a aria-current=\\"page\\" class=\\"is-active\\" href=\\"/\\">HVA ER?</a></li><li><a href=\\"/kontakt\\">KONTAKT</a></li><li><a href=\\"/admin\\">ADMIN</a></li></ul></nav><button>LOGG UT</button></header><div id=\\"home-page-container\\"><div id=\\"home-page-banner\\" class=\\"card\\"><div id=\\"home-page-banner-fade\\"><div id=\\"home-page-banner-content-wrapper\\"><h1>HOLOSEUM</h1><p>Et mobilt mini-museum som gir deg en smakebit av et ekte museum. Sammen med våre samarbeidspartnere har vi designet 3D-modeller av museumsgjenstander som kan vises hvor som helst!<br><br>Museumsgjenstandene vises som et 3D-hologram via et nettbrett. Velg blant de ulike muséene og få tilsendt det du trenger for å utforske norsk historie!<br><br>Bla ned for å se hvordan det fungerer.</p><a href=\\"/kontakt\\"><button id=\\"contact-btn\\" class=\\"ui button\\">KONTAKT OSS</button></a></div></div></div><div id=\\"guide-container\\"><div class=\\"guide-point card aos-init aos-animate\\" data-aos=\\"fade-up\\"><h2>1</h2><img src=\\"\\" alt=\\"Animasjon - guide steg 1\\"><p>Sett sammen stativet</p></div><div class=\\"guide-point card aos-init aos-animate\\" data-aos=\\"fade-up\\" data-aos-delay=\\"500\\"><h2>2</h2><img src=\\"\\" alt=\\"Animasjon - guide steg 2\\"><p>Bruk et nettbrett og skann en QR-kode på kartet i brosjyren</p></div><div class=\\"guide-point card aos-init aos-animate\\" data-aos=\\"fade-up\\" data-aos-delay=\\"1000\\"><h2>3</h2><img src=\\"\\" alt=\\"Animasjon - guide steg 3\\"><p>Legg nettbrettet på toppen av stativet, og hologrammet vil vises</p></div></div></div><footer></footer></div>"`;

exports[`HoloseumApp routes to NotFound component on invalid path 1`] = `"<div><div id=\\"not-found-container\\" class=\\"center-full-screen\\"><h1>404</h1><p>Siden finnes ikke</p><a href=\\"/\\">Klikk her for å gå til hovedsiden</a></div></div>"`;
